File: 05.括号生成/go/22.go
func generateParenthesis(n int) []string {
	var result []string
	if n == 0 {
		return result
	}
	dfs(n, n, "", &result)
	return result
}

func dfs(left int, right int, current string, result *[]string) {
	if left == 0 && right == 0 {
		//到底后写入
		*result = append(*result, current)
	}

	//左括号用完了
	if left > 0 {
		dfs(left-1, right, current+"(", result)
	}

	//右括号
	if right > left {
		dfs(left, right-1, current+")", result)
	}
}

File: 05.括号生成/cpp/22.cpp
class Solution {
public:
    void dfs(vector<string>& res, string str, int l, int r) {
        if (l < 0 || r < 0 || l > r) {
            return;
        }
        if (l == 0 && r == 0) {
            res.push_back(str);
        }
        dfs(res, str+"(", l-1, r);
        dfs(res, str+")", l, r-1);
    }
    
    vector<string> generateParenthesis(int n) {

        vector<string> res;
        dfs(res, "", n, n);
        return res;
    }
};
